From 98080a2e34eaf0085c3ea34079dfce89c10d0ee1 Mon Sep 17 00:00:00 2001
From: Andrey Gusakov <andrey.gusakov@cogentembedded.com>
Date: Mon, 30 May 2016 12:30:26 +0300
Subject: [PATCH] ASoC: PCM3168A: disable PM


Signed-off-by: Andrey Gusakov <andrey.gusakov@cogentembedded.com>
---
 sound/soc/codecs/pcm3168a.c |   12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/sound/soc/codecs/pcm3168a.c b/sound/soc/codecs/pcm3168a.c
index 8113566..2bee442 100644
--- a/sound/soc/codecs/pcm3168a.c
+++ b/sound/soc/codecs/pcm3168a.c
@@ -797,7 +797,8 @@ void pcm3168a_remove(struct device *dev)
 }
 EXPORT_SYMBOL_GPL(pcm3168a_remove);

-#ifdef CONFIG_PM
+
+#ifdef CONFIG_PM__
 static int pcm3168a_rt_resume(struct device *dev)
 {
 	struct pcm3168a_priv *pcm3168a = dev_get_drvdata(dev);
@@ -856,6 +857,15 @@ static int pcm3168a_rt_suspend(struct device *dev)

 	return 0;
 }
+#else
+static int pcm3168a_rt_resume(struct device *dev)
+{
+	return 0;
+}
+static int pcm3168a_rt_suspend(struct device *dev)
+{
+	return 0;
+}
 #endif

 const struct dev_pm_ops pcm3168a_pm_ops = {
--
1.7.10.4
