diff --git a/arch/arm/boot/dts/r8a7791-porter.dts b/arch/arm/boot/dts/r8a7791-porter.dts
index 30efd02..54cc27c 100644
--- a/arch/arm/boot/dts/r8a7791-porter.dts
+++ b/arch/arm/boot/dts/r8a7791-porter.dts
@@ -25,6 +25,19 @@
  */
 
 /dts-v1/;
+
+/* Reserve 1MB used as shared memory for ADSP */
+/memreserve/ 0x60000000 0x00100000;
+
+/* Reserve memory containing DSP code/data */
+/memreserve/ 0x62000000 0x00400000;
+
+/* Reserve memory containing DSP reset vectors */
+/memreserve/ 0x63000000 0x00001000;
+
+/* Reserve memory used by tracing facility */
+/memreserve/ 0x68000000 0x08000000;
+
 #include "r8a7791.dtsi"
 #include <dt-bindings/gpio/gpio.h>
 
@@ -37,7 +50,7 @@
 	};
 
 	chosen {
-		bootargs = "console=ttySC6,38400 ignore_loglevel rw root=/dev/nfs ip=dhcp vmalloc=384M";
+		bootargs = "console=ttySC10,38400 ignore_loglevel rw root=/dev/mmcblk0p1 ip=none vmalloc=384M";
 	};
 
 	memory@40000000 {
@@ -108,6 +121,8 @@
 	};
 
 	sound {
+		status = "disabled";
+
 		compatible = "simple-audio-card";
 
 		simple-audio-card,format = "i2s";
diff --git a/arch/arm/boot/dts/r8a7791.dtsi b/arch/arm/boot/dts/r8a7791.dtsi
index b10e656..d87af54 100644
--- a/arch/arm/boot/dts/r8a7791.dtsi
+++ b/arch/arm/boot/dts/r8a7791.dtsi
@@ -710,7 +710,7 @@
 			clocks = <&extal_clk>;
 			#clock-cells = <1>;
 			clock-output-names = "main", "pll0", "pll1", "pll3",
-					     "lb", "qspi", "sdh", "sd0", "z";
+					     "lb", "qspi", "sdh", "sd0", "adsp", "z";
 		};
 
 		/* Variable factor clocks */
@@ -964,13 +964,13 @@
 		mstp5_clks: mstp5_clks@e6150144 {
 			compatible = "renesas,r8a7791-mstp-clocks", "renesas,cpg-mstp-clocks";
 			reg = <0 0xe6150144 0 4>, <0 0xe615003c 0 4>;
-			clocks = <&zs_clk>, <&zs_clk>, <&extal_clk>, <&p_clk>;
+			clocks = <&zs_clk>, <&zs_clk>, <&cpg_clocks R8A7791_CLK_ADSP>, <&extal_clk>, <&p_clk>;
 			#clock-cells = <1>;
-			renesas,clock-indices = <R8A7791_CLK_AUDIO_DMAC0 R8A7791_CLK_AUDIO_DMAC1
+			renesas,clock-indices = <R8A7791_CLK_AUDIO_DMAC0 R8A7791_CLK_AUDIO_DMAC1 R8A7791_CLK_ADSP_MOD
 					R8A7791_CLK_THERMAL R8A7791_CLK_PWM>;
-			clock-indices = <R8A7791_CLK_AUDIO_DMAC0 R8A7791_CLK_AUDIO_DMAC1
+			clock-indices = <R8A7791_CLK_AUDIO_DMAC0 R8A7791_CLK_AUDIO_DMAC1 R8A7791_CLK_ADSP_MOD
 					R8A7791_CLK_THERMAL R8A7791_CLK_PWM>;
-			clock-output-names = "audmac0", "audmac1", "thermal", "pwm";
+			clock-output-names = "audmac0", "audmac1", "adsp_mod", "thermal", "pwm";
 		};
 		mstp7_clks: mstp7_clks@e615014c {
 			compatible = "renesas,r8a7791-mstp-clocks", "renesas,cpg-mstp-clocks";
@@ -1398,4 +1398,11 @@
 			ssi9: ssi@9 { interrupts = <0 379 IRQ_TYPE_LEVEL_HIGH>; };
 		};
 	};
+
+	adsp: adsp@ec800000 {
+		compatible = "renesas,r8a7791-adsp", "renesas,rcar-gen2-adsp";
+		reg = <0 0xec800000 0 0x160>, <0 0xec680000 0 0x1000>;
+		interrupts = <0 237 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&mstp5_clks R8A7791_CLK_ADSP_MOD>;
+	};
 };
diff --git a/arch/arm/mach-shmobile/board-porter-reference.c b/arch/arm/mach-shmobile/board-porter-reference.c
index a286a80..09986bf 100644
--- a/arch/arm/mach-shmobile/board-porter-reference.c
+++ b/arch/arm/mach-shmobile/board-porter-reference.c
@@ -133,6 +133,34 @@ static void __init porter_add_du_device(void)
 	platform_device_register_full(&info);
 }
 
+static struct asoc_simple_card_info rsnd_adsp_card_info = {
+	.name		= "ADSP-AK4643",
+	.codec		= "ak4642-codec.2-0012",
+	.platform	= "snd-soc-dummy",
+	.daifmt		= SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_CBM_CFM,
+	.cpu_dai = {
+		.name	= "snd-soc-dummy-dai",
+	},
+	.codec_dai = {
+		.name	= "ak4642-hifi",
+		.sysclk	= 11289600,
+	},
+};
+
+static void __init porter_add_rsnd_adsp_device(void)
+{
+	struct platform_device_info cardinfo = {
+		.parent         = &platform_bus,
+		.name           = "asoc-simple-card",
+		.id             = -1,
+		.data           = &rsnd_adsp_card_info,
+		.size_data      = sizeof(struct asoc_simple_card_info),
+		.dma_mask       = DMA_BIT_MASK(32),
+	};
+
+	platform_device_register_full(&cardinfo);
+}
+
 /*
  * This is a really crude hack to provide clkdev support to platform
  * devices until they get moved to DT.
@@ -909,6 +937,7 @@ static void __init porter_add_standard_devices(void)
 			     porter_auxdata_lookup, NULL);
 
 	porter_add_du_device();
+	porter_add_rsnd_adsp_device();
 	porter_add_usb_devices();
 	porter_add_camera0_device();
 #if IS_ENABLED(CONFIG_VIDEO_RENESAS_VSP1) && \
diff --git a/arch/arm/mach-shmobile/setup-rcar-gen2.c b/arch/arm/mach-shmobile/setup-rcar-gen2.c
index da16ebd..d164b81 100644
--- a/arch/arm/mach-shmobile/setup-rcar-gen2.c
+++ b/arch/arm/mach-shmobile/setup-rcar-gen2.c
@@ -172,9 +172,9 @@ void __init rcar_gen2_reserve(void)
 {
 	struct memory_reserve_config mrc;
 
-	/* reserve 256 MiB at the top of the physical legacy 32-bit space */
+	/* reserve 128 MiB at the top of the physical legacy 32-bit space */
 	memset(&mrc, 0, sizeof(mrc));
-	mrc.reserved = SZ_256M;
+	mrc.reserved = SZ_128M;
 
 	of_scan_flat_dt(rcar_gen2_scan_mem, &mrc);
 #ifdef CONFIG_DMA_CMA
diff --git a/drivers/clk/shmobile/clk-rcar-gen2.c b/drivers/clk/shmobile/clk-rcar-gen2.c
index 6bb7214..cee85a8 100644
--- a/drivers/clk/shmobile/clk-rcar-gen2.c
+++ b/drivers/clk/shmobile/clk-rcar-gen2.c
@@ -33,6 +33,7 @@ struct rcar_gen2_cpg {
 #define CPG_FRQCRC			0x000000e0
 #define CPG_FRQCRC_ZFC_MASK		(0x1f << 8)
 #define CPG_FRQCRC_ZFC_SHIFT		8
+#define CPG_ADSPCKCR			0x0000025c
 
 /* -----------------------------------------------------------------------------
  * Z Clock
@@ -207,6 +208,13 @@ static const struct clk_div_table cpg_sd01_div_table[] = {
 	{  0,  0 },
 };
 
+/* ADSP divisors */
+static const struct clk_div_table cpg_adsp_div_table[] = {
+	{  1,  3 }, {  2,  4 }, {  3,  6 }, {  4,  8 },
+	{  5, 12 }, {  6, 16 }, {  7, 18 }, {  8, 24 },
+	{ 10, 36 }, { 11, 48 }, {  0,  0 },
+};
+
 /* -----------------------------------------------------------------------------
  * Initialization
  */
@@ -263,6 +271,11 @@ rcar_gen2_cpg_register_clock(struct device_node *np, struct rcar_gen2_cpg *cpg,
 		shift = 0;
 	} else if (!strcmp(name, "z")) {
 		return cpg_z_clk_register(cpg);
+	} else if (!strcmp(name, "adsp")) {
+		parent_name = "pll1";
+		return clk_register_divider_table(NULL, name, parent_name, 0,
+						 cpg->reg + CPG_ADSPCKCR, 0,
+						 4, 0, cpg_adsp_div_table, &cpg->lock);
 	} else {
 		return ERR_PTR(-EINVAL);
 	}
diff --git a/include/dt-bindings/clock/r8a7791-clock.h b/include/dt-bindings/clock/r8a7791-clock.h
index 4c8b9d2..c768e9e 100644
--- a/include/dt-bindings/clock/r8a7791-clock.h
+++ b/include/dt-bindings/clock/r8a7791-clock.h
@@ -19,7 +19,8 @@
 #define R8A7791_CLK_QSPI		5
 #define R8A7791_CLK_SDH			6
 #define R8A7791_CLK_SD0			7
-#define R8A7791_CLK_Z			8
+#define R8A7791_CLK_ADSP		8
+#define R8A7791_CLK_Z			9
 
 /* MSTP0 */
 #define R8A7791_CLK_MSIOF0		0
@@ -71,6 +72,7 @@
 /* MSTP5 */
 #define R8A7791_CLK_AUDIO_DMAC1		1
 #define R8A7791_CLK_AUDIO_DMAC0		2
+#define R8A7791_CLK_ADSP_MOD		6
 #define R8A7791_CLK_THERMAL		22
 #define R8A7791_CLK_PWM			23
 
