diff --git a/arch/arm64/boot/dts/renesas/r8a7796.dtsi b/arch/arm64/boot/dts/renesas/r8a7796.dtsi
index afdd69d..1d814eb 100644
--- a/arch/arm64/boot/dts/renesas/r8a7796.dtsi	
+++ b/arch/arm64/boot/dts/renesas/r8a7796.dtsi
@@ -45,6 +45,41 @@
 		tsc2 = &tsc3;
 	};
 
+  reserved-memory {
+		#address-cells = <2>;
+		#size-cells = <2>;
+		ranges;
+
+		/* device specific region for Lossy Decompression */
+		lossy_decompress: linux,lossy_decompress {
+			no-map;
+			reg = <0x00000000 0x54000000 0x0 0x03000000>;
+		};
+
+		/* For Audio DSP */
+		adsp_reserved: linux,adsp {
+			compatible = "shared-dma-pool";
+			reusable;
+			reg = <0x00000000 0x57000000 0x0 0x01000000>;
+		};
+
+		/* global autoconfigured region for contiguous allocations */
+		linux,cma {
+			compatible = "shared-dma-pool";
+			reusable;
+			reg = <0x00000000 0x58000000 0x0 0x18000000>;
+			linux,cma-default;
+		};
+
+		/* device specific region for contiguous allocations */
+		mmp_reserved: linux,multimedia {
+			compatible = "shared-dma-pool";
+			reusable;
+			reg = <0x00000000 0x70000000 0x0 0x10000000>;
+		};
+	};
+
+
 	psci {
 		compatible = "arm,psci-1.0";
 		method = "smc";
@@ -2389,7 +2424,8 @@
 				      "audmac.1", "audmac.0",
 				      "adg";
 			power-domains = <&sysc R8A7796_PD_ALWAYS_ON>;
-			status = "disabled";
+      memory-region = <&adsp_reserved>;
+      status = "okay";
 		};
 
 		gsx: gsx@fd000000 {
