
diff --git a/../patch generator/a/ravb_main.c b/../patch generator/b/ravb_main.c
index 6eeeb4d..a3e558c 100644
--- a/drivers/net/ethernet/renesas/ravb_main.c
+++ b/drivers/net/ethernet/renesas/ravb_main.c
@@ -770,8 +770,6 @@ static int ravb_dmac_init(struct net_device *ndev, bool start)
 	if (start) {
 		/* Frame Receive */
 		ravb_write(ndev, 0x00000003, RIC0);
-		/* Receive FIFO full warning */
-		ravb_write(ndev, 0x80000000, RIC1);
 		/* Receive FIFO full error, Descriptor Empty */
 		ravb_write(ndev, 0x80000003, RIC2);
 		/* Frame Transmited, Timestamp FIFO updated */
@@ -2145,6 +2143,7 @@ static struct ravb_plat_data *ravb_parse_dt(struct device *dev,
 
 static const struct of_device_id ravb_match_table[] = {
 	{ .compatible = "renesas,gether-r8a7790", .data = &r8a779x_data_giga },
+	{ .compatible = "renesas,gether-r8a7791", .data = &r8a779x_data_giga },
 	{ .compatible = "renesas,gether-r8a7794", .data = &r8a779x_data_giga },
 	{ }
 };
